ConfigurationBinder source generator generates non-compilable code with new property and internal setter

`dotnet build` on the following projects results in the generated source from the Configuration.Binder source generator to not compile.

**Library**

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>net8.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
  </PropertyGroup>

</Project>
```

```C#
namespace ClassLibrary1;

public class CredentialDescription
{
    public string? Certificate { get; set; }
}

public class CertificateDescription : CredentialDescription
{
    public new string? Certificate
    {
        get
        {
            return base.Certificate;
        }
        protected internal set
        {
            base.Certificate = value;
        }
    }
}
```

**App**

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net8.0</TargetFramework>
    <Nullable>enable</Nullable>
    <ImplicitUsings>enable</ImplicitUsings>
    <EnableConfigurationBindingGenerator>true</EnableConfigurationBindingGenerator>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.Extensions.Hosting" Version="8.0.0" />
    <PackageReference Include="Microsoft.Extensions.Configuration.Binder" Version="8.0.1" />
  </ItemGroup>

  <ItemGroup>
    <ProjectReference Include="..\ClassLibrary1\ClassLibrary1.csproj" />
  </ItemGroup>
</Project>
```

```C#
using ClassLibrary1;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.Hosting;

var builder = Host.CreateApplicationBuilder(args);

builder.Configuration.Bind(new CertificateDescription());

var host = builder.Build();
host.Run();
```

### Expected result

I expect the code to compile. It may emit warnings if it was unable to bind properties correctly.

### Actual result

```
C:\Users\eerhardt\source\repos\WorkerService20\WorkerService20\obj\Debug\net8.0\Microsoft.Extensions.Configuration.Binder.SourceGeneration\Microsoft.
Extensions.Configuration.Binder.SourceGeneration.ConfigurationBindingGenerator\BindingExtensions.g.cs(65,17): error CS0272: The property or indexer '
CertificateDescription.Certificate' cannot be used in this context because the set accessor is inaccessible [C:\Users\eerhardt\source\repos\WorkerSer
vice20\WorkerService20\WorkerService20.csproj]
    0 Warning(s)
    1 Error(s)
```

### Generated code

```C#
// <auto-generated/>

#nullable enable annotations
#nullable disable warnings

// Suppress warnings about [Obsolete] member usage in generated code.
#pragma warning disable CS0612, CS0618

namespace System.Runtime.CompilerServices
{
    using System;
    using System.CodeDom.Compiler;

    [GeneratedCode("Microsoft.Extensions.Configuration.Binder.SourceGeneration", "8.0.9.7805")]
    [AttributeUsage(AttributeTargets.Method, AllowMultiple = true)]
    file sealed class InterceptsLocationAttribute : Attribute
    {
        public InterceptsLocationAttribute(string filePath, int line, int column)
        {
        }
    }
}

namespace Microsoft.Extensions.Configuration.Binder.SourceGeneration
{
    using Microsoft.Extensions.Configuration;
    using System;
    using System.CodeDom.Compiler;
    using System.Collections.Generic;
    using System.Globalization;
    using System.Runtime.CompilerServices;

    [GeneratedCode("Microsoft.Extensions.Configuration.Binder.SourceGeneration", "8.0.9.7805")]
    file static class BindingExtensions
    {
        #region IConfiguration extensions.
        /// <summary>Attempts to bind the given object instance to configuration values by matching property names against configuration keys recursively.</summary>
        [InterceptsLocation(@"C:\Users\eerhardt\source\repos\WorkerService20\WorkerService20\Program.cs", 7, 23)]
        public static void Bind_CertificateDescription(this IConfiguration configuration, object? instance)
        {
            if (configuration is null)
            {
                throw new ArgumentNullException(nameof(configuration));
            }

            if (instance is null)
            {
                return;
            }

            var typedObj = (global::ClassLibrary1.CertificateDescription)instance;
            BindCore(configuration, ref typedObj, defaultValueIfNotFound: false, binderOptions: null);
        }
        #endregion IConfiguration extensions.

        #region Core binding extensions.
        private readonly static Lazy<HashSet<string>> s_configKeys_CertificateDescription = new(() => new HashSet<string>(StringComparer.OrdinalIgnoreCase) { "Certificate" });

        public static void BindCore(IConfiguration configuration, ref global::ClassLibrary1.CertificateDescription instance, bool defaultValueIfNotFound, BinderOptions? binderOptions)
        {
            ValidateConfigurationKeys(typeof(global::ClassLibrary1.CertificateDescription), s_configKeys_CertificateDescription, configuration, binderOptions);

            if (configuration["Certificate"] is string value0)
            {
                instance.Certificate = value0;
            }
        }


        /// <summary>If required by the binder options, validates that there are no unknown keys in the input configuration object.</summary>
        public static void ValidateConfigurationKeys(Type type, Lazy<HashSet<string>> keys, IConfiguration configuration, BinderOptions? binderOptions)
        {
            if (binderOptions?.ErrorOnUnknownConfiguration is true)
            {
                List<string>? temp = null;
        
                foreach (IConfigurationSection section in configuration.GetChildren())
                {
                    if (!keys.Value.Contains(section.Key))
                    {
                        (temp ??= new List<string>()).Add($"'{section.Key}'");
                    }
                }
        
                if (temp is not null)
                {
                    throw new InvalidOperationException($"'ErrorOnUnknownConfiguration' was set on the provided BinderOptions, but the following properties were not found on the instance of {type}: {string.Join(", ", temp)}");
                }
            }
        }
        #endregion Core binding extensions.
    }
}
```

cc @ericstj @tarekgh @eiriktsarpalis

Ran some tests for comparison:

`protected internal`
```
C:\scratch\ee\app>dotnet build
Restore complete (1.2s)
You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy
  lib succeeded (2.6s) → C:\scratch\ee\lib\bin\Debug\net8.0\lib.dll
  app failed with errors (0.9s)
    C:\scratch\ee\app\obj\Debug\net8.0\Microsoft.Extensions.Configuration.Binder.SourceGeneration\Microsoft.Extensions.Configuration.Binder.SourceGeneration.ConfigurationBindingGenerator\BindingExtensions.g.cs(65,17): error CS0272: The property or indexer 'CertificateDescription.Certificate' cannot be used in this context because the set accessor is inaccessible [C:\scratch\ee\app\app.csproj]

Build failed with errors in 5.4s
```


`internal`
```
C:\scratch\ee\app>dotnet build
Restore complete (0.4s)
You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy
  lib succeeded (0.3s) → C:\scratch\ee\lib\bin\Debug\net8.0\lib.dll
  app failed with errors (0.2s)
    C:\scratch\ee\app\obj\Debug\net8.0\Microsoft.Extensions.Configuration.Binder.SourceGeneration\Microsoft.Extensions.Configuration.Binder.SourceGeneration.ConfigurationBindingGenerator\BindingExtensions.g.cs(65,17): error CS0200: Property or indexer 'CertificateDescription.Certificate' cannot be assigned to -- it is read only [C:\scratch\ee\app\app.csproj]

Build failed with errors in 1.3s
```
`private`
```
C:\scratch\ee\app>dotnet build
Restore complete (0.4s)
You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy
  lib succeeded (0.3s) → C:\scratch\ee\lib\bin\Debug\net8.0\lib.dll
  app failed with errors (0.2s)
    C:\scratch\ee\app\obj\Debug\net8.0\Microsoft.Extensions.Configuration.Binder.SourceGeneration\Microsoft.Extensions.Configuration.Binder.SourceGeneration.ConfigurationBindingGenerator\BindingExtensions.g.cs(65,17): error CS0200: Property or indexer 'CertificateDescription.Certificate' cannot be assigned to -- it is read only [C:\scratch\ee\app\app.csproj]
```

If I remove the `new string? Certificate` or change it's name to `new string? Certificate2` it builds successfully without warnings.

So the bug here happens to be when a derived type hides a property of a base and changes visibility.  Perhaps the source generator should be using the actual defining type of the member when accessing it?  I wonder if we could even share logic in a type heirarchy - where a Bind method bound only the members of that type and called another Bind call for the base - down the chain.

--- a/src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Parser.cs
+++ b/src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Parser.cs
@@ -660,6 +660,12 @@ namespace Microsoft.Extensions.Configuration.Binder.SourceGeneration
                         if (member is IPropertySymbol { IsIndexer: false, IsImplicitlyDeclared: false } property && !IsUnsupportedType(property.Type))
                         {
                             string propertyName = property.Name;
+
+                            if (properties?.ContainsKey(propertyName) is true)
+                            {
+                                continue;
+                            }
+
                             TypeRef propertyTypeRef = EnqueueTransitiveType(typeParseInfo, property.Type, DiagnosticDescriptors.PropertyNotSupported, propertyName);
 
                             AttributeData? attributeData = property.GetAttributes().FirstOrDefault(a => SymbolEqualityComparer.Default.Equals(a.AttributeClass, _typeSymbols.ConfigurationKeyNameAttribute));

--- a/src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Parser.cs
+++ b/src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Parser.cs
@@ -661,7 +661,7 @@ namespace Microsoft.Extensions.Configuration.Binder.SourceGeneration
                         {
                             string propertyName = property.Name;
 
-                            if (properties?.ContainsKey(propertyName) is true)
+                            if (property.IsOverride || properties?.ContainsKey(propertyName) is true)
                             {
                                 continue;
                             }

--- a/src/libraries/Microsoft.Extensions.Configuration.Binder/tests/Common/ConfigurationBinderTests.TestClasses.cs
+++ b/src/libraries/Microsoft.Extensions.Configuration.Binder/tests/Common/ConfigurationBinderTests.TestClasses.cs
@@ -928,24 +928,65 @@ namespace Microsoft.Extensions
         {
             public string A { get; set; }
             public string B { get; set; }
+        }
+
+        public class BaseForHiddenMembers
+        {
+            public string A { get; set; }
+            public string B { get; set; }
             public TestSettingsEnum E {get; set;}
+
+            public virtual string C { get => CBase; set => CBase = value; }
+
+            public string CBase;
+
+            public virtual string D { get; }
+
+            public virtual string F { get => FBase; set => FBase = value; }
+            public string FBase;
+
+
+            public virtual int X { get => XBase; set => XBase = value; }
+            public int XBase;
         }
 
         public enum TestSettingsEnum2
         {
-            OptionA = TestSettingsEnum.Option1,
-            OptionB = TestSettingsEnum.Option2,
+            // Note - the reflection binder will try to bind to every member
+            Option1 = TestSettingsEnum.Option1,
+            Option2 = TestSettingsEnum.Option2,
         }
 
-        public class DerivedClassWithHiddenMembers : SimplePoco
+        public class IntermediateDerivedClass : BaseForHiddenMembers
         {
-            public new string A { get; } = "Derived";
+            public new virtual string D { get => DBase; set => DBase = value; }
+            public string DBase;
+
+            public override string F { get => "IF"; }
+
+        }
+
+        public class DerivedClassWithHiddenMembers : IntermediateDerivedClass
+        {
+            public new string A { get; } = "ADerived";
             public new int B { get; set; }
             public new TestSettingsEnum2 E
             {
                 get => (TestSettingsEnum2)base.E;
                 set => base.E = (TestSettingsEnum)value;
             }
+
+            // only override get
+            public override string C { get => "DC"; }
+
+            // override new only get
+            public override string D { get => "DD"; }
+
+            // two overrides of only get
+            public override string F { get => "DF"; }
+
+            // override only set
+            public override int X { set => base.X = value + 1; }
         }
 
     }

--- a/src/libraries/Microsoft.Extensions.Configuration.Binder/tests/Common/ConfigurationBinderTests.TestClasses.cs
+++ b/src/libraries/Microsoft.Extensions.Configuration.Binder/tests/Common/ConfigurationBinderTests.TestClasses.cs
@@ -928,6 +928,24 @@ namespace Microsoft.Extensions
         {
             public string A { get; set; }
             public string B { get; set; }
+            public TestSettingsEnum E {get; set;}
+        }
+
+        public enum TestSettingsEnum2
+        {
+            OptionA = TestSettingsEnum.Option1,
+            OptionB = TestSettingsEnum.Option2,
+        }
+
+        public class DerivedClassWithHiddenMembers : SimplePoco
+        {
+            public new string A { get; } = "Derived";
+            public new int B { get; set; }
+            public new TestSettingsEnum2 E
+            {
+                get => (TestSettingsEnum2)base.E;
+                set => base.E = (TestSettingsEnum)value;
+            }
         }
 
     }

--- a/src/libraries/Microsoft.Extensions.Configuration.Binder/tests/Common/ConfigurationBinderTests.cs
+++ b/src/libraries/Microsoft.Extensions.Configuration.Binder/tests/Common/ConfigurationBinderTests.cs
@@ -2475,15 +2475,63 @@ if (!System.Diagnostics.Debugger.IsAttached) { System.Diagnostics.Debugger.Launc
         [Fact]
         public void CanBindToClassWithNewProperties()
         {
-            var config = TestHelpers.GetConfigurationFromJsonString(@"{""A"":""Should not bind"", ""B"":""5"", ""E"":""OptionB""}");
+            /// the source generator will bind to the most derived property only.
+            /// the reflection binder will bind the same data to all properties (including hidden).
+            
+            var config = TestHelpers.GetConfigurationFromJsonString("""
+                {
+                    "A": "AVal",
+                    "B": "5",
+                    "C": "CVal",
+                    "D": "DVal",
+                    "E": "Option2",
+                    "F": "FVal",
+                    "X": "52"
+                }
+                """);
             var obj = new DerivedClassWithHiddenMembers();
-            string initialA = obj.A;
 
             config.Bind(obj);
 
-            Assert.Equal(initialA, obj.A);
+            BaseForHiddenMembers baseObj = obj;
+            IntermediateDerivedClass intermediateObj = obj;
+
+            Assert.Equal("ADerived", obj.A);
+#if BUILDING_SOURCE_GENERATOR_TESTS
+            // source generator will not set hidden property
+            Assert.Null(baseObj.A);
+#else
+            // reflection binder will set hidden property
+            Assert.Equal("AVal", baseObj.A);
+#endif
+
             Assert.Equal(5, obj.B);
-            Assert.Equal(TestSettingsEnum2.OptionB, obj.E);
+#if BUILDING_SOURCE_GENERATOR_TESTS
+            // source generator will not set hidden property
+            Assert.Null(baseObj.B);
+#else
+            // reflection binder will set hidden property
+            Assert.Equal("5", baseObj.B);
+#endif
+
+            Assert.Equal(TestSettingsEnum2.Option2, obj.E);
+            Assert.Equal(TestSettingsEnum.Option2, baseObj.E);
+
+            Assert.Equal("DC", obj.C);
+            // The setter should still be called, even when only getter is overridden.
+            Assert.Equal("CVal", obj.CBase);
+
+            // can hide a readonly property with r/w property
+            Assert.Null(baseObj.D);
+            Assert.Equal("DD", obj.D);
+            // The setter should still be called, even when only getter is overridden.
+            Assert.Equal("DVal", obj.DBase);
+
+            Assert.Equal("DF", obj.F);
+            Assert.Equal("FVal", obj.FBase);
+
+            Assert.Equal(53, obj.X);
+            Assert.Equal(53, obj.XBase);
         }
     }
 }

--- a/src/libraries/Microsoft.Extensions.Configuration.Binder/tests/Common/ConfigurationBinderTests.cs
+++ b/src/libraries/Microsoft.Extensions.Configuration.Binder/tests/Common/ConfigurationBinderTests.cs
@@ -2471,5 +2471,19 @@ if (!System.Diagnostics.Debugger.IsAttached) { System.Diagnostics.Debugger.Launc
             IConfigurationSection IConfiguration.GetSection(string key) =>
                 this[key] is null ? null : new ConfigurationSection(this, key);
         }
+
+        [Fact]
+        public void CanBindToClassWithNewProperties()
+        {
+            var config = TestHelpers.GetConfigurationFromJsonString(@"{""A"":""Should not bind"", ""B"":""5"", ""E"":""OptionB""}");
+            var obj = new DerivedClassWithHiddenMembers();
+            string initialA = obj.A;
+
+            config.Bind(obj);
+
+            Assert.Equal(initialA, obj.A);
+            Assert.Equal(5, obj.B);
+            Assert.Equal(TestSettingsEnum2.OptionB, obj.E);
+        }
     }
 }

